### PR TITLE
[fix] #1742: Concrete errors returned in `core` instructions.

### LINKED ISSUE #1742: Concretise Errors in `iroha_core`
In Iroha core, many error types were implicitly converted into `eyre::Report`s which replaced transaction rejection reason with `Some other error happened`.

### PR DESCRIPTION
**Pull request by appetrosyan**
Signed-off-by: Aleksandr <a-p-petrosyan@yandex.ru>

### Description of the Change

Opaque `eyre::Report`s have been replaced with specific `Error` enums. 

### Issue

Closes #1742 

### Benefits

Opaque error types that could only have been printed are now also possible to handle and match. 

### Possible Drawbacks

None

### CODE REVIEW

File: data_model/src/fixed.rs [hunk 1]
```diff
@@ -98,15 +98,29 @@ pub enum FixedPointOperationError {
...
     Conversion(#[source] ConvertError),
-    /// The arithmetic operation failed.
-    #[error("Arithmetic error")]
-    Arithmetic(#[source] ArithmeticError),
+    /// Overflow
+    #[error("Overflow")]
+    Overflow,
+    /// Division by zero
+    #[error("Division by zero")]
+    DivZero,
```
Comment (s8sato) [inline data_model/src/fixed.rs:106]:
```suggestion
    DivisionByZero,
```

File: data_model/src/fixed.rs [hunk 2]
```diff
@@ -98,15 +98,29 @@ pub enum FixedPointOperationError {
...
+    /// Domain violation. E.g. computing `sqrt(-1)`
+    #[error("Domain violation")]
+    Domain,
```
Comment (s8sato) [inline data_model/src/fixed.rs:109]:
```suggestion
    DomainViolation,
```

**Review by s8sato — DISMISSED** [review]

File: core/src/smartcontracts/isi/asset.rs [hunk 3]
```diff
@@ -294,15 +304,17 @@ pub mod isi {
 
 /// Asset-related query implementations.
 pub mod query {
-    use eyre::{eyre, Result, WrapErr};
+    use eyre::{Result, WrapErr};
+    // use futures::TryFutureExt;
```
Comment (eadventurous) [inline core/src/smartcontracts/isi/asset.rs:308]: Maybe you forgot to remove it?

**Review by s8sato — APPROVED** [review]

**Comment by mversic** [comment]
can we protect ourselves from this happening in the future?

**Review by mversic — APPROVED** [review]

**Review by eadventurous — APPROVED** [review]

**Pull request merged**